File: Daily/198. House Robber/Solution.cpp
#include <iostream>
#include <vector>
#include <bits/stdc++.h>
using namespace std;
class Solution {
    public:
        int rob(vector<int>& nums) {
            /*
                Unoptimized Solution using recursion
                T(n-0) = T(n-1) + T(n-2)
                T(n-1) = T(n-2) + T(n-3)
                T(n-2) = T(n-3) + T(n-4)
                ........................
                T(n-n) = 0
                Time Complexity = O(2^n)
                Space Complexity = O(n)
            */
            function<int(int)> find_Max_Amount = [&](int index) {
                if(index >= nums.size()) return 0;
                int choice1 = nums[index] + find_Max_Amount(index + 2);
                int choice2 = find_Max_Amount(index + 1);
                return max(choice1,choice2);
            };
            return find_Max_Amount(0);
        }
};
int main() {
    int n; cin >> n;
    vector<int> nums(n, 0); for(int index = 0; index < n; index++) cin >> nums[index];
    Solution sol; int ans = sol.rob(nums);
    cout << ans << endl;
}

File: Daily/198. House Robber/Solution.py
from typing import List
class Solution:
    def rob(self, nums: List[int]) -> int:
        '''
            Unoptimized Solution
            Time Complexity = O(2^n)
            Space Complexity = O(n)
        '''
        def find_Max_Amount(index: int = 0) -> int:
            if index >= len(nums): return 0
            return max(nums[index] + find_Max_Amount(index + 2), find_Max_Amount(index + 1))
        return find_Max_Amount()
if __name__ == '__main__':
    nums = [int(x) for x in input().split()]; sol = Solution()
    print(sol.rob(nums))